### PR TITLE
Added DNAnexus app files

### PR DESCRIPTION
**Pull request by mcrusch**
Files were placed in new directory dx_app.  Added the four dx-specific
files (2x readmes, dxapp.json, entry point), as well as a build.sh
script, which performs a docker build, docker save, and then dx build.

### CODE REVIEW

File: dx_app/dxapp.json [hunk 1]
```diff
@@ -0,0 +1,134 @@
+{
+  "name": "stjude_xenocp",
+  "title": "XenoCP (St. Jude)",
+  "summary": "Xenograft sample cleaning pipeline",
+  "dxapi": "1.0.1",
+  "version": "3.0.0",
+  "openSource": false,
+  "developers": [ "user-lding", "user-mrusch", "user-athrasher" ],
```
Comment (zaeleus) [inline dx_app/dxapp.json]:
```suggestion
  "developers": ["user-lding", "user-mrusch", "user-athrasher"],
```

File: dx_app/src/stjude_xenocp.sh [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+ref_db_prefix: $container_reference_dir/$reference_prefix
+output_prefix: $output_prefix
+output_extension: $output_extension
+EOF
+
+    echo "  [*] Loading container image ..."
+    image_tarfile_path=/stjude/xenocp-docker.tar
+    if [ -e $image_tarfile_path.gz ]
+    then gunzip $image_tarfile_path.gz
+    fi
```
Comment (zaeleus) [inline dx_app/src/stjude_xenocp.sh]: Given `/stjude/xenocp-docker.tar.gz` is a static asset, why check if it's compressed or not?

File: dx_app/src/stjude_xenocp.sh [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        echo "A custom host genome was not provided but Host Genome was set to custom." >&2
+        exit 1
+      fi
+      dx download -o $local_reference_dir $ref_bwa_index_tgz
+      (
+        cd $local_reference_dir
+        tar xzf *
+      )
+    fi
+    reference_prefix=`cd $local_reference_dir; echo *.bwt | sed 's/.bwt$//'`
```
Comment (zaeleus) [inline dx_app/src/stjude_xenocp.sh]: Prefer the `$()` syntax over backticks for command substitution. [See rationale](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_xcu_chap02.html#tag_23_02_06_03).

File: dx_app/src/stjude_xenocp.sh [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    docker load -i $image_tarfile_path
+    
+    echo "=== Execution ==="
+    
+    # Don't make assumptions about the tag that was used when the image was
+    # built, other than that it should be "xenocp".  Since this is run in a
+    # clean environment, and we only did a single docker load, the method
+    # below should be a safe way to determine the image.
+    echo "  [*] Determine image ID ..."
+    image_id=`docker images -q xenocp | head -n 1`
```
Comment (zaeleus) [inline dx_app/src/stjude_xenocp.sh]: Prefer the `$()` syntax over backticks for command substitution. [See rationale](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_xcu_chap02.html#tag_23_02_06_03).

File: dx_app/src/stjude_xenocp.sh [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        tar xzf *
+      )
+    fi
+    reference_prefix=`cd $local_reference_dir; echo *.bwt | sed 's/.bwt$//'`
+    if ! ls $local_reference_dir/$reference_prefix*
+    then echo "Problem obtaining bwa index" >&2; exit 1
+    fi
+
+    echo "  [*] Printing instance information ..."
+    cat /etc/lsb-release
```
Comment (zaeleus) [inline dx_app/src/stjude_xenocp.sh]: Prefer `/etc/os-release`, which is a more adopted standard (systemd) than LSB compliance.

File: dx_app/Readme.md [hunk 6]
```diff
@@ -0,0 +1,20 @@
+<!-- dx-header -->
+# XenoCP: Xenograft sample cleaning pipeline
+
```
[the remark this answers is not in the document]
Comment (mcrusch) [reply dx_app/Readme.md]: What do these dx-header comments actually do?

File: dx_app/build.sh [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  (
+    cd ..
+    docker build -t xenocp:$VERS .
+  )
+fi
+
+mkdir -p resources/stjude
+docker save xenocp:$VERS | gzip -c > resources/stjude/xenocp-docker.tar.gz
+
+dx build $DX_BUILD_ARGS
```
[the remark this answers is not in the document]
Comment (mcrusch) [reply dx_app/build.sh]: Can you explain more?

File: dx_app/dxapp.json [hunk 8]
```diff
@@ -0,0 +1,134 @@
+{
+  "name": "stjude_xenocp",
+  "title": "XenoCP (St. Jude)",
+  "summary": "Xenograft sample cleaning pipeline",
+  "dxapi": "1.0.1",
```
[the remark this answers is not in the document]
Comment (mcrusch) [reply dx_app/dxapp.json]: I'm not sure.  This line is unedited from the file I imported.  Suggestions on how to set?

File: dx_app/dxapp.json [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+  "version": "3.0.0",
+  "openSource": false,
```
[the remark this answers is not in the document]
Comment (mcrusch) [reply dx_app/dxapp.json]: Unedited from imported version.  I will update.

File: dx_app/dxapp.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (zaeleus) [reply dx_app/dxapp.json]: I'm not aware of a 1.0.1. This should be set to `1.0.0`, the default used by `dx-app-wizard`: https://github.com/dnanexus/dx-toolkit/blob/c0110e48d5e92c79ece7480ed2d9c6b853a2dd32/src/python/dxpy/scripts/dx_app_wizard.py#L46

File: dx_app/src/stjude_xenocp.sh [hunk 10]
```diff
@@ -0,0 +1,136 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (mcrusch) [reply dx_app/src/stjude_xenocp.sh]: Updated, pending commit.

File: dx_app/build.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (zaeleus) [reply dx_app/build.sh]: When dealing with positional parameters, you typically don't want to interpret or expand them when passing them to another command. `$@` is won't do this, while setting it another variable does. E.g.,

```bash
#!/usr/bin/env bash
ruby -e 'p ARGV' -- "$@"
BUILD_ARGS="$@"
ruby -e 'p ARGV' -- $BUILD_ARGS
```

```
$ ./params.sh foo "bar baz"
["foo", "bar baz"]
["foo", "bar", "baz"]
```

Note how the argument "bar baz" is no longer passed as a single argument.

File: dx_app/src/stjude_xenocp.sh (hunk 3 above)
Comment (mcrusch) [reply dx_app/src/stjude_xenocp.sh]: I think backticks are still more readable for most developers and so I will continue to use backticks for simple command substitution, relegating `$()` for to "new applications that nest command substitutions or attempt to embed complex scripts." (from the rationale link)

File: dx_app/src/stjude_xenocp.sh (hunk 5 above)
Comment (mcrusch) [reply dx_app/src/stjude_xenocp.sh]: Updated, pending commit.

File: dx_app/src/stjude_xenocp.sh (hunk 2 above)
Comment (mcrusch) [reply dx_app/src/stjude_xenocp.sh]: This way the entry point is agnostic wrt to the compression state of the image.  docker build spits out plain tar, and we're performing an extra step to gzip compress it.  I could imagine in the future deciding not to do the extra compression/decompression, though for now it's the best option based on my evaluation.  So, I prefer leaving these couple of lines in to keep the flexibility at the entry-point level.

File: dx_app/Readme.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (zaeleus) [reply dx_app/Readme.md]: It's parsed by DNAnexus when presenting the readme on their site. It's usually just the title and summary.

File: dx_app/Readme.developer.md [hunk 11]
```diff
@@ -0,0 +1,7 @@
+# XenoCP Developer Information
+
+Source is available at https://github.com/adamdingliang/XenoCP
```
Comment (adthrasher) [inline dx_app/Readme.developer.md]:
```suggestion
Source is available at https://github.com/stjude/XenoCP
```

**Pull request merged**